### PR TITLE
Datawrapper changes

### PR DESCRIPTION
**Pull request by emdash-ie**
## What does this change?

Some changes are needed to this script to make it work correctly for datawrapper graphics. Because I don't want to change how the script works for other consumers, my plan is to keep these commits off of main, marked by a version tag specifically for datawrapper graphics.

This PR documents the changes for the purposes of code review, but I don't plan to merge it.

## How to test

I followed this process to test (and will repeat to get screenshots for the PR):

- Hosted the version of the script from this branch at https://interactive.guim.co.uk/datawrapper-test/iframeMessenger.js
- Updated [a test datawrapper graphic's embed](https://interactive.guim.co.uk/datawrapper-test/embed/jCxow/4/) to replace the usual version of iframe-messenger with that one
- Embedded this within an article in composer-CODE and viewed it [in preview](https://preview.code.dev-gutools.co.uk/money/2024/nov/18/ghg)

## How can we measure success?

No more overflow and visible scrollbars in datawrapper graphics!

## Have we considered potential risks?

Since this removes the usual `overflow: hidden` added to the `html` element by this script, if Datawrapper make any changes that cause graphics to overflow those elements then the overflow will make it into the rest of the page. (Visually there seems to be a small overflow at the moment, but it's less than one pixel and can be accommodated easily by the margin we have below the graphics.)

### Before

Screenshot 2024-12-13 at 14 36 39

### After

Screenshot 2024-12-13 at 14 44 12

### CODE REVIEW

File: src/iframeMessenger.js [hunk 1]
```diff
@@ -154,7 +154,7 @@
          * @return {int} Height integer
          */
         function _getHeight() {
-            var height = parseInt(document.body.offsetHeight, 10);
+            var height = Math.ceil(document.body.getBoundingClientRect().height) + 1;
```
Comment (jonathonherbert) [inline src/iframeMessenger.js:158]: Why +1? Perhaps we could name this number with a variable to avoid a comment.

**Review by jonathonherbert — APPROVED** [review]
This makes sense! The test article displays correctly for me. If there's anything else I can do to help verification, let me know. One non-blocking comment to capture information for future devs.

If we did want to merge these changes in future, perhaps we could contain this path in an option we pass to `enableAutoResize` — option name is a hard one, perhaps `useUpdatedHeightCalculation` (unsure how you'd describe the difference in methodology concisely!)

Perhaps this might also lead to versioning concerns for previous datawrapper graphics, as this script has a single, unversioned deployment. I suppose we'd need to update the datawrapper deployment script to include this new argument, and then redeploy every datawrapper interactive to receive the new code?

**Comment by emdash-ie** [comment]
> If we did want to merge these changes in future, perhaps we could contain this path in an option we pass to `enableAutoResize` — option name is a hard one, perhaps `useUpdatedHeightCalculation` (unsure how you'd describe the difference in methodology concisely!)

Actually, the updated height calculation is the one part of this that I think I would be happy extending to the existing version of the script, at least partially. Since the change to using `getBoundingClientRect()` is about improving precision (to fractions of a pixel), I would expect it to be more correct in all cases (though I’m not 100% sure I’m confident enough in my knowledge to make such a strong claim).

(However, the 1px added is just to accommodate Datawrapper graphics. They appear to have a small amount of overflow in them when used with this script, even with these updates, for reasons I don’t fully understand yet.)

> Perhaps this might also lead to versioning concerns for previous datawrapper graphics, as this script has a single, unversioned deployment. I suppose we'd need to update the datawrapper deployment script to include this new argument, and then redeploy every datawrapper interactive to receive the new code?

Yes, unfortunately so.

File: src/iframeMessenger.js (hunk 1 above)
Comment (emdash-ie) [reply src/iframeMessenger.js:158]: Unfortunately there’s no good name I can come up with for the 1 here – by testing I’ve determined that when I leave it out there’s overflow, whereas when I include it there isn’t (and I have no better understanding for why it’s necessary). Adding a comment sounds good though!

**Comment by emdash-ie** [comment]
Now the PR has been reviewed, I have added a tag to the tip of this branch sharing a name with the version number in the script: https://github.com/guardian/iframe-messenger/releases/tag/0.2.10-datawrapper

I’ll close this PR now :)

**Comment by emdash-ie** [comment]
These changes have worked! After merging https://github.com/guardian/datawrapper-import/pull/37, I republished https://app.datawrapper.de/edit/aFsY0/publish, which is used in https://www.theguardian.com/environment/2024/nov/20/the-climate-crisis-in-charts-how-2024-has-set-unwanted-new-records, and the overflow has been successfully removed.

Before:

Screenshot 2024-12-19 at 16 05 42

After:

Screenshot 2024-12-19 at 16 07 59

**Pull request closed without merging**